contracts/StorageFees.cdc

```diff
@@ -0,0 +1,180 @@
+                    self.purchases.append(StorageCapacityPurchase(storageCapacity:purchase.storageCapacity- storageCapacity,  flowCost: purchase.flowCost - relativeFlow))
+                    break;
+                }
+            }
+
+            if storageCapacity != UInt64(0) {
+                panic("Cannot refund storage")
+            }
+
+            return <- self.vault.withdraw(flowAmount)
+        }
+
+        init(storageCapacity: UInt64, address: Address, vault: @FlowToken.Vault) {
+            self.id = StorageFees.idCounter
+            StorageFees.idCounter = StorageFees.idCounter + UInt64(1)
+            self.storageCapacity = storageCapacity
+            self.address = address
+            self.purchases = [ StorageCapacityPurchase(storageCapacity: storageCapacity, flowCost: vault.balance)]
+            self.vault <- vault
+        }
+
+        destroy() {
+            // in general the transaction will be reverted because the account doesnt have any capacity anymore
+            destroy self.vault
+        }
+    }
+
+    pub fun setupStorageForAccount(paymentVault: @FlowToken.Vault, authAccount: AuthAccount){
+        pre{
+            paymentVault.balance != StorageCapacity.flowPerAccountCreation: "Account creation cost exactly ".concat(StorageCapacity.flowPerAccountCreation).concat(" Flow tokens.")
+            authAccount.check<&StorageCapacity{StorageCapacityCapability}>(/public/storageCapacity): "Account already has storage setup"
+        }
+        let storageCapacity <- create StorageCapacity(
+            storageCapacity: StorageCapacity.minimumAccountStorage,
+            address: authAccount.address,
+            vault: <- paymentVault)
+
+        authAccount.save(<- storageCapacity, to: /storage/storageCapacity)
+
+        authAccount.link<&StorageCapacity{StorageCapacityCapability}>(
+            /public/storageCapacity,
+            target: /storage/storageCapacity
+        )
+    }
+
+    pub fun addStorageCapacity(to: Address, amount: UInt64, paymentVault: @FlowToken.Vault){
+        pre{
+            amount % StorageFees.minimumStorageUnit == UInt64(0): "Amount of storage capacity to add must be a multiple of the minimum storage unit"
+            paymentVault.balance != StorageFees.flowPerByte * UFix64(amount): "Adding ".concat(amount.toString()).concat(" storage capacity cost exactly ").concat(StorageCapacity.flowPerByte*amount).concat(" Flow tokens.")
+            
+        }
+        let storageCapacityCapability = getAccount(to).getCapability<&StorageCapacity{StorageCapacityCapability}>(/public/storageCapacity)!.borrow()
+        if storageCapacityCapability == nil {
+            panic("Account needs to be setup first")
+        }
+
+        storageCapacityCapability?.addStorageCapacity(address: to, amount: amount, paymentVault: paymentVault)
+    }
+
+    pub fun refundStorageCapacity(from: Address, storageCapacityReference: &StorageCapacity, storageAmount: UInt64): @FlowToken.Vault {
```

Comment:
I'm not sure why we need an Address argument here. The StorageCapacity object has everything it needs to know which account it's managing. In fact, I'm not sure why this method exists at all, given that it just calls a method on one of the arguments!

Reply:
One of the problems Im trying to prevent is storage swapping: two accounts swapping storage because one account bought it when it was cheap, and the other wants it. 

The process is as follows:
1. Account A has a lot of storage
2. in one transaction: A moves its StorageCapacity away from the standard location and calls `setupStorageForAccount` to get a new one so it has some capacity and the Tx doesn't fail 
3. it can now transfer its original storage to account B and B swaps it with its current storage 

I was trying to add an account field to StorageCapacity to prevent B using(or refunding) StorageCapacity created for A.

That is also what I was trying to do with the idCounter you commented above... I still need to think this scenario through...